README.md
# GImage

[![Build Status](https://github.com/joseluisq/gimage/actions/workflows/devel.yml/badge.svg?branch=master)](https://github.com/joseluisq/gimage/actions/workflows/devel.yml) [![](https://poser.pugx.org/joseluisq/gimage/version)](https://packagist.org/packages/joseluisq/gimage) [![Latest Unstable Version](https://poser.pugx.org/joseluisq/gimage/v/unstable)](//packagist.org/packages/joseluisq/gimage) [![Total Downloads](https://poser.pugx.org/joseluisq/gimage/downloads)](https://packagist.org/packages/joseluisq/gimage) [![License](https://poser.pugx.org/joseluisq/gimage/license)](https://packagist.org/packages/joseluisq/gimage)

> A PHP library for easy image handling. 🖼

__GImage__ is a simple and small library based on [PHP Image Processing and GD](http://php.net/manual/en/book.image.php) for processing images easily.

![A simple presentation card with GImage](https://cloud.githubusercontent.com/assets/1700322/18941713/eed7fa34-85d8-11e6-8033-bf787e4aa236.png)

_Presentation card built with GImage - [View code example](https://joseluisq.github.io/gimage/examples/creating-a-presentation-card/)_

## Features

- Load an image from a local path, URL or image resource.
- Create shapes such as rectangles or ellipses with opacity.
- Resize, scale or crop images proportionally.
- Rotate images, shapes or texts.
- Embed text with custom TTF fonts.
- Compose a pool of images with `Canvas`.
- Swap image formats such as JPEG, PNG or GIF.
- Save images locally or output them on the browser.
- Save several copies of the same image.
- Render an image in memory and return its resource.

## Requirements

GImage requires PHP [`7.4`](https://www.php.net/releases/7_4_0.php) or a recent PHP version with the latest [GD Extension](http://php.net/manual/en/book.image.php).

## Install

```sh
composer require joseluisq/gimage
```

## Usage

Load an external PNG image and save it as JPG:

```php
<?php

use GImage\Image;

// PNG image (600x199)
$url = 'https://i.imgur.com/G5MR088.png';

$arch = new Image();
$arch
    // Load from URL
    ->load($url)
    // Scale to 50% (300x99)
    ->scale(0.5)
    // Change the format to JPG
    ->toJPG()
    // Saving in local path
    ->save('arch.jpg');
```

See [GImage Website](https://bit.ly/gimage-php) for detailed usage instructions and more advanced examples.

## Changelog

Check out the [CHANGELOG.md](./CHANGELOG.md) file.

## Examples

See [code examples](./examples) directory.

## Development

### Run tests

Via [Composer](https://getcomposer.org/)

```sh
composer install
composer run-script test
```

Via [Docker](https://www.docker.com/) (multi PHP versions test)

```sh
make docker-tests
```

## Contribution

Feel free to send some [Pull request](https://github.com/joseluisq/gimage/pulls) or file an [issue](https://github.com/joseluisq/gimage/issues).

## License

This work is primarily distributed under the terms of the [MIT license](LICENSE-MIT).

©2015-present [Jose Quintana](https://joseluisq.net)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]


## [4.0.0] - 2022-10-24

This major release is fundamentally an upgrade of the minimum PHP version required (`7.4.x` or newer), API remains basically the same, new posibility to load an image from a `resource` or `GdImage` input, two new small getter methods for GImage along with other notable project improvements.
And as noted, the project resumes its maintenance after a long period. 

### Added
- Annotate methods that return same class via return [`static`](https://wiki.php.net/rfc/static_return_type).
- PHP `8.0` support. PR [#35](https://github.com/joseluisq/gimage/issues/35)
- PHP `8.1` support. PR [#37](https://github.com/joseluisq/gimage/issues/37)
  - Improve PHP `8.1` tests and CI
  - Fixes missing types casting for PHP `8.1`
- Load an image from a `resource` (PHP 7.4) or [`\GdImage`](https://php.watch/versions/8.0/gdimage) (PHP `8.x`) via `GImage->load($src)` method. PR [#38](https://github.com/joseluisq/gimage/issues/38)
- New `GImage->isImageResource()` (PHP 7.4) and `GImage->isImageGdImage()` (PHP `8.x`) methods.

### Deprecated
- PHP `7.3` or lower is deprecated. Now GImage requires PHP `7.4.x` or `8.x` along with a latest GD extension.

### Changed
- Improve example files.
- GitHub Actions as a new CI for testing the library against PHP `7.4.x`, `8.0.x` and `8.1.x`.
- Several documentation improvements.

## [3.0.6] - 2018-04-26

### Added
- `isImageString` method. (PR [#29](https://github.com/joseluisq/gimage/issues/30))

### Fixed
- `$url` is undefined in loadImageFromResource method. (PR [#30](https://github.com/joseluisq/gimage/issues/30) by @franklee0902)
- `Utils::isJPGResource` and `Utils::isPNGResource` method require a string parameter. (PR [#30](https://github.com/joseluisq/gimage/issues/30) by @franklee0902)

## [3.0.5] - 2018-04-24

### Added
- Load an image resource directly. (Issue [#29](https://github.com/joseluisq/gimage/issues/29))

## [3.0.4] - 2017-12-23

### Fixed
- Undefined variable `$info`. (PR [#26](https://github.com/joseluisq/gimage/pull/26) by @sunnyphp)
- First example syntax. (PR [#25](https://github.com/joseluisq/gimage/pull/25) by @sunnyphp)


## [3.0.3] - 2017-10-07

### Added
- __Image:__ Added `render()` method can render the image in-memory and return the resource. (REF [02c5480](https://github.com/joseluisq/gimage/commit/02c5480))
- __Image:__ Added `render()` example about at `examples/render.php`

### Changed
- __Text:__ Default value in `setContent($str =  '')` function.

### Fixed
- Fixed issue #20.
- Doc comments in classes.

## [3.0.2] - 2017-04-30

### Added
- __Image:__ Added the `Image::addOpacityFilter()` method with protected visiblity.

### Changed
- Remove unused imports. (PR [#16](https://github.com/joseluisq/gimage/pull/16) by @matiit)
- Simplify importing Image properties in the `Image::from()` method. (PR [#14](https://github.com/joseluisq/gimage/pull/14) by @matiit)

### Fixed
- __Figure:__ Fixes for add correctly the opacity filter in rectangles and ellipses. (Issue [#15](https://github.com/joseluisq/gimage/issues/15))
- __Image:__ Render fixes for add correctly PNG's opacity.
- __Utils:__ Fixed the `Utils::fixPNGOpacity()` method.


## [3.0.1] - 2017-04-14

### Added
- `Utils::fixPNGOpacity()` function that fix the opacity value between `0` and `1` for PNG alpha value.

### Changed
- `setOpacity` (`Figure`, `Image` and `Text` classes) method now support values from `0` to `1` only.

```php
<?php
// Set opacity to 50%
$figure->setOpacity(0.5)
```

### Fixed
- `setOpacity` now works if it loads some JPEG image and then saves it as PNG.
- Fixed the opacity tests.
- Fixed the examples for opacity support.

### Removed
- Some unnecessary functions for figure class.

### Deprecated
- `setOpacity` doesn't support values from 0 to 100.


## [3.0.0] - 2017-02-16
### Added
- Feature: `isEllipse()` method support in `Figure` for create ellipses.
- Feature: `isRectangle()` method support in `Figure` for create rectangles.
- `CHANGELOG.md` file.

### Changed
- `Figure` elements are `PNG` by default.

src/Image.php
<?php
/*
 * This file is part of GImage.
 *
 * (c) Jose Quintana <https://joseluisq.net>
 *
 * This source file is subject to the MIT license that is bundled
 * with this source code in the file LICENSE.
 */

namespace GImage;

/**
 * A simple extended GD class for easy image handling.
 * This is the parent class for Figure and Canvas.
 *
 * @author Jose Quintana <http://git.io/joseluisq>
 *
 * @property string   $name
 * @property string   $filename
 * @property int      $width
 * @property int      $height
 * @property int      $x
 * @property int      $y
 * @property int      $boxWidth
 * @property int      $boxHeight
 * @property int      $boxX
 * @property int      $boxY
 * @property int      $type              IMAGETYPE_JPEG
 * @property string   $extension         Default 'jpg'
 * @property resource $resource|\GdImage
 * @property int      $quality
 * @property int      $opacity           From 0 to 1
 * @property string   $from              Default 'local'
 * @property bool     $preserve          Default FALSE
 * @property string   $mimetype          Default 'image/jpeg'
 */
class Image
{
    protected $name;
    protected $filename;
    protected $width     = 0;
    protected $height    = 0;
    protected $x         = 0;
    protected $y         = 0;
    protected $boxWidth  = 0;
    protected $boxHeight = 0;
    protected $boxX      = 0;
    protected $boxY      = 0;
    protected $type      = IMAGETYPE_JPEG;
    protected $extension = 'jpg';
    protected $resource;
    protected $quality  = 100;
    protected $opacity  = 1;
    protected $from     = 'local';
    protected $preserve = false;
    protected $mimetype = 'image/jpeg';

    /**
     * Loads an image from Image or Figure class.
     *
     * @param Image $element image or Figure class
     */
    public function __construct($element = null)
    {
        $this->from($element);
    }

    /**
     * Loads an image from Image or Figure class.
     *
     * @param Image $element image or Figure class
     *
     * @return \GImage\Image|static
     */
    public function from($element = null)
    {
        if (!empty($element) && $element instanceof Image) {
            foreach (get_object_vars($element) as $key => $value) {
                $this->$key = $value;
            }

            return $this;
        }
    }

    /**
     * Loads an image from a local path, external url or image string.
     *
     * @param string $src local path, external url or image string
     *
     * @return \GImage\Image|static
     */
    public function load($src)
    {
        if (empty($src)) {
            return $this;
        }

        // 1. resource or \GdImage
        if (is_resource($src) || $src instanceof \GdImage) {
            $this->loadImageFromResource($src);
        // 2. string image
        } elseif (is_string($src) && $this->isImageStringByString($src)) {
            $this->loadImageFromString($src);
        // 3. string url image path
        } elseif (filter_var($src, FILTER_VALIDATE_URL)) {
            $this->loadImageFromURL($src);
        // 4. string file path
        } elseif (is_file($src)) {
            $this->loadImageFromFile($src);
        }

        return $this;
    }

    /**
     * Load an image from URL.
     *
     * @param string $url Image URL
     *
     * @return void
     */
    private function loadImageFromURL($url)
    {
        $image = $this->fetchImageContentFromURL($url);

        if (empty($image)) {
            return;
        }

        if (Utils::isJPGResource($image)) {
            $this->filename  = $url;
            $this->extension = 'jpg';
            $this->type      = IMAGETYPE_JPEG;
        }

        if (Utils::isPNGResource($image)) {
            $this->filename  = $url;
            $this->extension = 'png';
            $this->type      = IMAGETYPE_PNG;
        }

        $this->from     = 'external';
        $this->resource = imagecreatefromstring($image);
        $this->width    = $this->boxWidth    = imagesx($this->resource);
        $this->height   = $this->boxHeight   = imagesy($this->resource);
    }

    /**
     * Fetch an image string content from URL.
     *
     * @param string $url image URL
     *
     * @return string String data
     */
    private function fetchImageContentFromURL($url)
    {
        $data = null;

        if ($stream = fopen($url, 'r')) {
            $data = stream_get_contents($stream);
            fclose($stream);
        }

        return $data;
    }

    /**
     * Load an image from a resource or `\GdImage`.
     *
     * @param string $resource image resource or `\GdImage`
     *
     * @return void
     */
    private function loadImageFromResource($resource)
    {
        $this->extension = 'png';
        $this->type      = IMAGETYPE_PNG;
        $this->from      = is_resource($resource) ? 'resource' : 'gdimage';
        $this->resource  = $resource;
        $this->width     = $this->boxWidth     = imagesx($this->resource);
        $this->height    = $this->boxHeight    = imagesy($this->resource);
    }

    /**
     * Load an image from image string.
     *
     * @param string $imagestring Image string
     *
     * @return void
     */
    private function loadImageFromString($imagestring)
    {
        if (Utils::isJPGResource($imagestring)) {
            $this->extension = 'jpg';
            $this->type      = IMAGETYPE_JPEG;
        }

        if (Utils::isPNGResource($imagestring)) {
            $this->extension = 'png';
            $this->type      = IMAGETYPE_PNG;
        }

        $this->from     = 'imagestring';
        $this->resource = imagecreatefromstring($imagestring);
        $this->width    = $this->boxWidth    = imagesx($this->resource);
        $this->height   = $this->boxHeight   = imagesy($this->resource);
    }

    /**
     * Checks if string is Image string.
     *
     * @param string $imagestring image string
     *
     * @return bool
     */
    private function isImageStringByString($imagestring)
    {
        return Utils::isJPGResource($imagestring) || Utils::isPNGResource($imagestring);
    }

    /**
     * Load an image file from path.
     *
     * @param string $filepath File path
     *
     * @return void
     */
    private function loadImageFromFile($filepath)
    {
        $this->from     = 'local';
        $this->filename = $filepath;

        if (!is_readable($filepath)) {
            return;
        }

        $image                            = null;
        $extension                        = '';
        list($width, $height, $imageType) = getimagesize($filepath);

        switch ($imageType) {
            case IMAGETYPE_GIF:
                $extension = 'gif';
                $image     = imagecreatefromgif($filepath);
                break;

            case IMAGETYPE_PNG:
                $extension = 'png';
                $image     = imagecreatefrompng($filepath);
                break;

            case IMAGETYPE_JPEG:
                $extension = 'jpg';
                $image     = imagecreatefromjpeg($filepath);
                break;
        }

        if ($image) {
            $this->type      = $imageType;
            $this->resource  = $image;
            $this->extension = $extension;
            $this->width     = $this->boxWidth     = $width;
            $this->height    = $this->boxHeight    = $height;
        }
    }

    /**
     * Gets the resource or `\GdImage` object (PHP 8.0+) of the image.
     *
     * @return resource|\GdImage
     */
    public function getResource()
    {
        return $this->resource;
    }

    /**
     * Gets filename path of the image.
     *
     * @return string
     */
    public function getFilename()
    {
        return $this->filename;
    }

    /**
     * Gets the image file name.
     *
     * @return string
     */
    public function getName()
    {
        return $this->name;
    }

    /**
     * Gets the image type code (IMAGETYPE_JPEG, IMAGETYPE_PNG or IMAGETYPE_GIF).
     *
     * @return int
     */
    public function getType()
    {
        return $this->type;
    }

    /**
     * Gets extension (jpg, png or gif).
     *
     * @return string
     */
    public function getExtension()
    {
        return $this->extension;
    }

    /**
     * Gets quality.
     *
     * @return int
     */
    public function getQuality()
    {
        return $this->quality;
    }

    /**
     * Gets opacity.
     *
     * @return int
     */
    public function getOpacity()
    {
        return $this->opacity;
    }

    /**
     * Gets width.
     *
     * @return int
     */
    public function getWidth()
    {
        return $this->width;
    }

    /**
     * Gets height.
     *
     * @return int
     */
    public function getHeight()
    {
        return $this->height;
    }

    /**
     * Gets top position.
     *
     * @return int
     */
    public function getTop()
    {
        return $this->y;
    }

    /**
     * Gets left position.
     *
     * @return int
     */
    public function getLeft()
    {
        return $this->x;
    }

    /**
     * Gets box left position.
     *
     * @return int
     */
    public function getBoxLeft()
    {
        return $this->boxX;
    }

    /**
     * Gets box top position.
     *
     * @return int
     */
    public function getBoxTop()
    {
        return $this->boxY;
    }

    /**
     * Gets box width.
     *
     * @return int
     */
    public function getBoxWidth()
    {
        return $this->boxWidth;
    }

    /**
     * Gets box height.
     *
     * @return int
     */
    public function getBoxHeight()
    {
        return $this->boxHeight;
    }

    /**
     * Gets where the image was loaded: 'local' or 'external'.
     *
     * @return string
     */
    public function getLoadedFrom()
    {
        return $this->from;
    }

    /**
     * Sets quality for image.
     *
     * @param int $quality Quality value from 0 to 100
     *
     * @return \GImage\Image|static
     */
    public function setQuality($quality)
    {
        $this->quality = $quality;

        return $this;
    }

    /**
     * Sets opacity.
     *
     * @param int $opacity Opacity value from 0 to 127
     *
     * @return \GImage\Image|static
     */
    public function setOpacity($opacity)
    {
        $this->opacity = $opacity > 127 ? 127 : $opacity;

        return $this;
    }

    /**
     * Sets image resource.
     *
     * @param resource $resource resource
     *
     * @return \GImage\Image|static
     */
    public function setResource($resource)
    {
        $this->resource = $resource;

        return $this;
    }

    /**
     * Sets left position of image.
     *
     * @param int $x left
     *
     * @return \GImage\Image|static
     */
    public function setLeft($x)
    {
        $this->x = $x;

        return $this;
    }

    /**
     * Sets top position of image.
     *
     * @param int $y top
     *
     * @return \GImage\Image|static
     */
    public function setTop($y)
    {
        $this->y = $y;

        return $this;
    }

    /**
     * Sets width of box image.
     *
     * @param int $width width
     *
     * @return \GImage\Image|static
     */
    public function setBoxWidth($width)
    {
        $this->boxWidth = $width;

        return $this;
    }

    /**
     * Sets height of box image.
     *
     * @param int $height height
     *
     * @return \GImage\Image|static
     */
    public function setBoxHeight($height)
    {
        $this->boxHeight = $height;

        return $this;
    }

    /**
     * Sets left position of box image.
     *
     * @param int $x left position
     *
     * @return \GImage\Image|static
     */
    public function setBoxLeft($x)
    {
        $this->boxX = $x * (-1);

        return $this;
    }

    /**
     * Sets top position of box image.
     *
     * @param int $y top position
     *
     * @return \GImage\Image|static
     */
    public function setBoxTop($y)
    {
        $this->boxY = $y * (-1);

        return $this;
    }

    /**
     * Checks if image was loaded from local path.
     *
     * @return bool
     */
    public function isLocal()
    {
        return $this->from == 'local';
    }

    /**
     * Checks if image was loaded from external url.
     *
     * @return bool
     */
    public function isExternal()
    {
        return $this->from == 'external';
    }

    /**
     * Checks if image was loaded from image string.
     *
     * @return bool
     */
    public function isImageString()
    {
        return $this->from == 'imagestring';
    }

    /**
     * Checks if image was loaded from an image resource (PHP 7.4).
     *
     * @return bool
     */
    public function isImageResource()
    {
        return $this->from == 'resource';
    }

    /**
     * Checks if image was loaded from a `\GdImage` object (PHP 8.0+).
     *
     * @return bool
     */
    public function isImageGdImage()
    {
        return $this->from == 'gdimage';
    }

    /**
     * Checks if image is a JPG.
     *
     * @return bool
     */
    public function isJPG()
    {
        return $this->extension == 'jpg';
    }

    /**
     * Checks if image is a PNG.
     *
     * @return bool
     */
    public function isPNG()
    {
        return $this->extension == 'png';
    }

    /**
     * Checks if image is a GIF.
     *
     * @return bool
     */
    public function isGIF()
    {
        return $this->extension == 'gif';
    }

    /**
     * Changes output format to JPG.
     *
     * @return \GImage\Image|static
     */
    public function toJPG()
    {
        $this->extension = 'jpg';
        $this->type      = IMAGETYPE_JPEG;
        $this->mimetype  = Utils::getMimetypeByImageType(IMAGETYPE_JPEG);

        return $this;
    }

    /**
     * Changes output format to PNG.
     *
     * @return \GImage\Image|static
     */
    public function toPNG()
    {
        $this->extension = 'png';
        $this->type      = IMAGETYPE_PNG;
        $this->mimetype  = Utils::getMimetypeByImageType(IMAGETYPE_PNG);

        return $this;
    }

    /**
     * Changes output format to GIF.
     *
     * @return \GImage\Image|static
     */
    public function toGIF()
    {
        $this->extension = 'gif';
        $this->type      = IMAGETYPE_GIF;
        $this->mimetype  = Utils::getMimetypeByImageType(IMAGETYPE_GIF);

        return $this;
    }

    /**
     * Preserves the resource image when save or output function is called.
     *
     * @param bool $preserve if it's true will preserve the resource image
     *
     * @return \GImage\Image|static
     */
    public function preserve($preserve = true)
    {
        $this->preserve = $preserve;

        return $this;
    }

    /**
     * Resize image proportionally basing on the height of the image.
     *
     * @param int $height
     *
     * @return \GImage\Image|static
     */
    public function resizeToHeight($height)
    {
        $width = $this->getPropWidth($height);
        $this->resize($width, $height);

        return $this;
    }

    /**
     * Resize image proportionally basing on the width of the image.
     *
     * @param int $width
     *
     * @return \GImage\Image|static
     */
    public function resizeToWidth($width)
    {
        $height = $this->getPropHeight($width);
        $this->resize($width, $height);

        return $this;
    }

    /**
     * Gets proportional width of image from height value.
     *
     * @param int $height
     *
     * @return int
     */
    public function getPropWidth($height)
    {
        $ratio = $height / $this->height;

        return (int) floor($this->width * $ratio);
    }

    /**
     * Gets proportional height of image from width value.
     *
     * @param int $width
     *
     * @return int
     */
    public function getPropHeight($width)
    {
        $ratio = $width / $this->width;

        return (int) floor($this->height * $ratio);
    }

    /**
     * Scales the image.
     *
     * @param int|float $scale
     *
     * @return \GImage\Image|static
     */
    public function scale($scale = 1)
    {
        $scale = $scale > 1 ? 1 : $scale;
        $scale = $scale < 0 ? 0 : $scale;

        $width  = (int) floor($this->width * $scale);
        $height = (int) floor($this->height * $scale);

        $this->resize($width, $height);

        return $this;
    }

    /**
     * Rotate the image with a given angle.
     *
     * @param int $angle
     *
     * @return \GImage\Image|static
     */
    public function rotate($angle = 0)
    {
        if ($this->resource) {
            $this->resource = imagerotate($this->resource, $angle, 0);
        }

        return $this;
    }

    /**
     * Cuts the image proportionally and centered.
     *
     * @param int $width  width crop
     * @param int $height height crop
     *
     * @return \GImage\Image|static
     */
    public function centerCrop($width, $height)
    {
        $pwidth  = $this->getPropWidth($height);
        $pheight = $this->getPropHeight($width);

        if ($pwidth == $width && $pheight == $height) {
            $this->resizeToWidth($width);
        } else {
            if ($pheight > $height) {
                $this->resizeToWidth($width);
            } else {
                $pheight += $height - $pheight;
                $this->resizeToHeight($pheight);
            }

            $x = (int) floor(($this->width - $width) / 2);
            $y = (int) floor(($this->height - $height) / 2);
            $this->crop($width, $height, $x, $y);
        }

        return $this;
    }

    /**
     * Cuts part of image.
     *
     * @param int $width  width crop
     * @param int $height height crop
     * @param int $x1     [Optional] x-coordinate of source point
     * @param int $y1     [Optional] y-coordinate of source point
     * @param int $dstX   [Optional] x-coordinate of destination point
     * @param int $dstY   [Optional] y-coordinate of destination point
     *
     * @return \GImage\Image|static
     */
    public function crop($width, $height, $x1 = 0, $y1 = 0, $dstX = 0, $dstY = 0)
    {
        $this->resize($width, $height, $x1, $y1, $dstX, $dstY, true);

        return $this;
    }

    /**
     * Resizes the image.
     *
     * @param int  $width  image's width
     * @param int  $height image's height
     * @param int  $x1     [Optional] Left position
     * @param int  $y1     [Optional] Top position
     * @param int  $dstX   [Optional] x-coordinate of destination point
     * @param int  $dstY   [Optional] y-coordinate of destination point
     * @param bool $isCrop [Optional] if it's true resize function will crop the image
     *
     * @return \GImage\Image|static
     */
    private function resize($width, $height, $x1 = 0, $y1 = 0, $dstX = 0, $dstY = 0, $isCrop = false)
    {
        if ($this->resource && $width > 0 && $height > 0) {
            $image = imagecreatetruecolor($width, $height);

            if ($this->isPNG()) {
                imagealphablending($image, false);
                imagesavealpha($image, true);
            }

            imagecopyresampled(
                $image,
                $this->resource,
                $dstX,
                $dstY,
                $x1,
                $y1,
                $width,
                $height,
                $isCrop ? $width : $this->width,
                $isCrop ? $height : $this->height
            );

            $this->resource = $image;
            $this->width    = $this->boxWidth    = imagesx($this->resource);
            $this->height   = $this->boxHeight   = imagesy($this->resource);
        }

        return $this;
    }

    /**
     * Saves the image to specific path.
     *
     * @param string $filename if it's null save function will save the image
     *                         in load path for default
     *
     * @return \GImage\Image|static
     */
    public function save($filename = null)
    {
        return $this->outputBuffer($filename);
    }

    /**
     * Outputs the image on browser.
     *
     * @return \GImage\Image|static
     */
    public function output()
    {
        return $this->outputBuffer(null, true);
    }

    /**
     * Render the image in-memory and return the resource.
     *
     * @return resource | null  Return the resource or null
     */
    public function render()
    {
        $image = null;

        if ($this->resource) {
            ob_start();
            $this->outputBufferByImage(null, $this->quality);
            $string = ob_get_contents();
            ob_end_clean();

            if (!empty($string)) {
                $image = imagecreatefromstring($string);
            }
        }

        return $image;
    }

    /**
     * Output the image to either the browser or a file.
     *
     * @param string $filename [Optional] Path to save image
     * @param bool   $output   [Optional] true to output the image
     *
     * @return \GImage\Image|static | resource
     */
    private function outputBuffer($filename = null, $output = false)
    {
        if (!$this->resource) {
            return $this;
        }

        if ($output) {
            $filename = null;
        } elseif (empty($filename) && $this->isLocal()) {
            $filename = $this->filename;
        } elseif (empty($filename) && $this->isExternal()) {
            $filename = $this->name;
        }

        if ($output) {
            header('Content-type: ' . $this->mimetype);
            ob_start();
        }

        $this->outputBufferByImage($filename, $this->quality);

        if ($output) {
            ob_end_flush();
        }

        if (!$this->preserve) {
            $this->destroy();
        }

        return $this;
    }

    private function outputBufferByImage($filename, $quality)
    {
        if ($this->isJPG()) {
            imagejpeg($this->resource, $filename, $quality);
        }

        if ($this->isPNG()) {
            if ($quality > 10) {
                $quality = 0;
            }

            imagesavealpha($this->resource, true);

            if (!is_subclass_of($this, Image::class)) {
                $this->addOpacityFilter();
            }

            imagepng($this->resource, $filename, $quality);
        }

        if ($this->isGIF()) {
            imagegif($this->resource, $filename);
        }
    }

    /**
     * Destroys the current resource.
     *
     * @return \GImage\Image|static
     */
    public function destroy()
    {
        if ($this->resource) {
            imagedestroy($this->resource);
        }

        $this->resource = null;

        return $this;
    }

    /**
     * Add opacity filter to the current resource.
     *
     * @return void
     */
    protected function addOpacityFilter()
    {
        $opacity = Utils::fixPNGOpacity($this->opacity);

        if ($opacity >= 0 && $opacity < 127) {
            imagealphablending($this->resource, false);
            imagefilter($this->resource, IMG_FILTER_COLORIZE, 0, 0, 0, $opacity);
        }
    }
}
